### ISSUE TITLE
CA1062 false positive for nullability analysis

### ISSUE THREAD

**Issue by bartdesmet** [dotnet/roslyn#64033]
**Version Used**: 

Microsoft Visual Studio Enterprise 2022
Version 17.2.4
VisualStudio.17.Release/17.2.4+32602.215
C# Tools   4.2.0-4.22281.5+8d3180e5f00d42f0f0295165f756f368f0cbfa44

**Steps to Reproduce**:

```xml
<!-- repro.csproj -->
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net6.0</TargetFrameworks>
    <ImplicitUsings>enable</ImplicitUsings>
    <TreatWarningsAsErrors>True</TreatWarningsAsErrors>
    <AnalysisLevel>latest-all</AnalysisLevel>
    <Nullable>enable</Nullable>
  </PropertyGroup>

</Project>
```

and

```csharp
// Program.cs
#define REPRO

using System.Collections.ObjectModel;
using System.Runtime.CompilerServices;

[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1014:MarkAssembliesWithClsCompliant")]

class Program { static void Main() { } }

namespace Bar
{
    public static class Foo
    {
        public static void Bar(IEnumerable<int>? xs)
        {
            var list = Qux(xs);

            _ = list.Count;
#if REPRO
            /*
            _ = list.Count;
                ~~~~
            CA1062  In externally visible method 'void Foo.Bar(IEnumerable<int>? xs)', validate parameter 'xs' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.
            */
#endif
        }

        public static ReadOnlyCollection<T> Qux<T>(IEnumerable<T>? enumerable)
        {
#if REPRO
            if (enumerable is ReadOnlyCollection<T> roc)
            {
                return roc;
            }
#endif

            return new ReadOnlyCollectionBuilder<T>(enumerable ?? Array.Empty<T>()).ToReadOnlyCollection();
        }
    }
}
```

**Expected Behavior**:

When commenting out `#define REPRO`, no CA1062 warning is produced complaining about `xs` being possibly null, even though the error is reported on `list` which is the result of calling `xs.ToReadOnly()` and has a non-nullable return type. It is unclear why there's spooky interaction with the *implementation* of this method even though its return type should suffice to reason over the result being non-null.

**Actual Behavior**:

```cpp
            _ = list.Count;
```

produces error

```
CA1062  In externally visible method 'void Foo.Bar(IEnumerable<int>? xs)', validate parameter 'xs' is non-null before using it. If appropriate, throw an ArgumentNullException when the argument is null or add a Code Contract precondition asserting non-null argument.
```

mentioning `xs`.

**Comment by bartdesmet** [comment]
FWIW, the real implementation of `Qux` is a more complex `ToReadOnly` extension method:

```csharp
public static ReadOnlyCollection<T> ToReadOnly<T>(this IEnumerable<T>? enumerable)
{
    if (enumerable == null)
    {
        return EmptyReadOnlyCollection<T>.Instance;
    }

    if (enumerable is ReadOnlyCollectionBuilder<T> builder)
    {
        return builder.ToReadOnlyCollection();
    }

    if (enumerable is ReadOnlyCollection<T> roc && IsTrueReadOnlyCollection(roc))
    {
        return roc;
    }

    return new ReadOnlyCollectionBuilder<T>(enumerable).ToReadOnlyCollection();
}
```

but I've simplified the repro by weeding out what seems to be the remaining `if` statement that changes the warning behavior. At the caller side, the warning is always reported on the first potentially null-dereferencing access on a local variable holding the non-null `ReadOnlyCollection<T>`, and the warning refers to the name of the parameter that was fed into `ToReadOnly`.
